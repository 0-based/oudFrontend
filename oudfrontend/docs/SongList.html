<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: SongList</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: SongList</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>SongList<span class="signature">(recieved, track, tracks, pause, resume, addToQueue, clickedItemId)</span><span class="type-signature"></span></h2>
        
            <div class="class-description">this is a component that renders the list of songs in playlists, albums, likedSongs or a "loading" animation if the songs are not recieved</div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="SongList"><span class="type-signature"></span>new SongList<span class="signature">(recieved, track, tracks, pause, resume, addToQueue, clickedItemId)</span><span class="type-signature"></span></h4>
    

    











    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>recieved</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            

            

            <td class="description last">the number of songs in the playlists, albums, likedSongs</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>track</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last">song object
                <h6>Properties</h6>
                

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>_id</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">track id</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>name</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">track title</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>albumId</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the id of the track's album</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>type</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the type of music that the track belongs to</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>audioUrl</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the URL of the track</td>
        </tr>

    
    </tbody>
</table>

            </td>
        </tr>

    

        <tr>
            
                <td class="name"><code>artist._id</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the Id of the singer</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>artist.name</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">name of the singer</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>artist.type</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the type of music that the artist produces</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>artist.image</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Image of the artist</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>track.artists</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;artist></span>


            
            </td>

            

            

            <td class="description last">the artists that produced the track</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>tracks</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;track></span>


            
            </td>

            

            

            <td class="description last">array of songs to be rendered</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>pause</code></td>
            

            <td class="type">
            
                
<span class="param-type">func</span>


            
            </td>

            

            

            <td class="description last">pauses the player</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>resume</code></td>
            

            <td class="type">
            
                
<span class="param-type">func</span>


            
            </td>

            

            

            <td class="description last">resumes the player</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>addToQueue</code></td>
            

            <td class="type">
            
                
<span class="param-type">func</span>


            
            </td>

            

            

            <td class="description last">adds array of tracks to queue</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>clickedItemId</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">used to mark all songs as unclicked</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    
    <dt class="tag-author">Author:</dt>
    <dd class="tag-author">
        <ul>
            <li><a href="mailto:ahmedwa1999@gmail.com">Ahmed Walid</a></li>
        </ul>
    </dd>
    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="commonComponents_songList.jsx.html">commonComponents/songList.jsx</a>, <a href="commonComponents_songList.jsx.html#line41">line 41</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        


    





    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="componentWillReceiveProps"><span class="type-signature"></span>componentWillReceiveProps<span class="signature">(nextProps)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    if the recieved props is changed it sets the state to the new props
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>nextProps</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last"></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="commonComponents_songList.jsx.html">commonComponents/songList.jsx</a>, <a href="commonComponents_songList.jsx.html#line62">line 62</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
            

    

    
    <h4 class="name" id="handleClick"><span class="type-signature"></span>handleClick<span class="signature">(id)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    it sets the state tothe new ID so it is sent to all the songs
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>id</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the id of the calling song</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="commonComponents_songList.jsx.html">commonComponents/songList.jsx</a>, <a href="commonComponents_songList.jsx.html#line55">line 55</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
            

    

    
    <h4 class="name" id="handlePlay"><span class="type-signature"></span>handlePlay<span class="signature">(id)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    if the calling song is playing it pauses/resums the playerotherwise, it calls handleQueue
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>id</code></td>
            

            <td class="type">
            
                
<span class="param-type">sttring</span>


            
            </td>

            

            

            <td class="description last">the id of the calling song</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="commonComponents_songList.jsx.html">commonComponents/songList.jsx</a>, <a href="commonComponents_songList.jsx.html#line87">line 87</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
            

    

    
    <h4 class="name" id="handleQueue"><span class="type-signature"></span>handleQueue<span class="signature">(id)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    it adds all the songs after the calling song to the queue
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>id</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">the id of the calling song</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="commonComponents_songList.jsx.html">commonComponents/songList.jsx</a>, <a href="commonComponents_songList.jsx.html#line71">line 71</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="addToPlaylist.html">addToPlaylist</a></li><li><a href="Album.html">Album</a></li><li><a href="CreatePlaylist.html">CreatePlaylist</a></li><li><a href="LikedSongs.html">LikedSongs</a></li><li><a href="Playlist.html">Playlist</a></li><li><a href="Song.html">Song</a></li><li><a href="SongList.html">SongList</a></li></ul><h3>Global</h3><ul><li><a href="global.html#addToQueue">addToQueue</a></li><li><a href="global.html#HeaderBody">HeaderBody</a></li><li><a href="global.html#HeaderBodyBottom">HeaderBodyBottom</a></li><li><a href="global.html#HeaderBodyTop">HeaderBodyTop</a></li><li><a href="global.html#pause">pause</a></li><li><a href="global.html#resume">resume</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 3.6.3</a> on Thu Apr 02 2020 22:33:34 GMT+0200 (Eastern European Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>