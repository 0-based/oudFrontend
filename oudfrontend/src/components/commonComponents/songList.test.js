import React from 'React'
import SongList from './songList';
import Enzyme, {shallow} from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import checkPropTypes from 'check-prop-types'
import renderer from 'react-test-renderer';

Enzyme.configure({adapter: new EnzymeAdapter()});

const fullProps = {
    recieved : true,
    tracks : [
        {"id": "19",
    "name": "إنسان",
    "artists": [
        {
            "id": "5",
            "name": "حمزة نمرة",
            "type": "shaaby",
            "image": "string"
        },
        {
            "id": "2",
            "name": "Sardeena",
            "type": "Shaaby",
            "image": "string"
        }
        ],
        "albumId": "1",
        "type": "Jazz",
        "audioUrl": "www.Facebook.com"
    },
    {
        "id": "20",
        "name": "سلاح التلميذ",
        "artists": [
        {
            "id": "3",
            "name": "ًWegz el wench",
            "type": "Trap",
            "image": "string"
        },
        {
            "id": "4",
            "name": "Amr Diab",
            "type": "farafeery",
            "image": "string"
        }
        ],
        "albumId": "1",
        "type": "Trap",
        "audioUrl": "www.facebook.com"
  }],
    pause : jest.fn(),
    resume : jest.fn(),
    addToQueue : jest.fn(),
    clickedItemId : '0'
}
const setup = (props={}) =>{
    return shallow(<SongList {...props}/>);
}

const findByTestAttr=(wrapper, val)=>{
    return wrapper.find(`[data-testid="${val}"]`);
}
describe('SongList Component', ()=>{

    
    describe('testing SongList Component with recieved equal true and tracks is empty',()=>{
        let component;
        let props = {recieved:true, tracks:[]}
        beforeEach (()=>{
            component = setup(props);
        })
        it('render songsList correctly when recived = true and tracks is empty', ()=>{
            const wrapper = findByTestAttr(component, 'songsList');
            expect(wrapper.length).toBe(1);
        });
        it("doesn't render songElement when recived = true and tracks is empty", ()=>{
            const wrapper = findByTestAttr(component, 'songElement');
            expect(wrapper.length).toBe(0);
        });
    });
    
    describe('testing SongList Component with recieved equal true and tracks not empty',()=>{
        let component;
        beforeEach (()=>{
            component = setup(fullProps);
        })
        it('render songsList correctly when recived = true and tracks not empty', ()=>{
            const wrapper = findByTestAttr(component, 'songsList');
            expect(wrapper.length).toBe(1);
        });
        it("doesn't render songElement when recived = true and tracks is not empty", ()=>{
            const wrapper = findByTestAttr(component, 'songElement');
            expect(wrapper.length).toBe(2);
        });
    });
    describe('checking propTypes', ()=>{
        const props10 ={recieved:true, tracks:{}}
        const props11 ={recieved:true, tracks:[]}
        const props01 ={recieved:4, tracks:[]}
        const props00 ={recieved:4, tracks:{}}
        //testing with full props
        it('should not throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {...fullProps}, 'prop', SongList.name);
            expect(result).toBeUndefined();
        }) 

        //testing pause function by sending invalid data
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {pause:'1'}, 'prop', SongList.name);
            expect(result).toBeDefined();
        }) 
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {pause:1}, 'prop', SongList.name);
            expect(result).toBeDefined();
        })
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {pause:true}, 'prop', SongList.name);
            expect(result).toBeDefined();
        })
        
        //testing resume function by sending invalid data
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {resume:'1'}, 'prop', SongList.name);
            expect(result).toBeDefined();
        }) 
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {resume:1}, 'prop', SongList.name);
            expect(result).toBeDefined();
        })
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {resume:true}, 'prop', SongList.name);
            expect(result).toBeDefined();
        })

        //testing addToQueue function by sending invalid data
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {addToQueue:'1'}, 'prop', SongList.name);
            expect(result).toBeDefined();
        }) 
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {addToQueue:1}, 'prop', SongList.name);
            expect(result).toBeDefined();
        })
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {addToQueue:true}, 'prop', SongList.name);
            expect(result).toBeDefined();
        })

        //testing clicked Item Id by sending false data
        it('should not throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, {clickedItemId:1}, 'prop', SongList.name);
            expect(result).toBeDefined();
        });


        it('should not throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, props11, 'prop', SongList.name);
            expect(result).toBeUndefined();
        });
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, props10, 'prop', SongList.name);
            console.log(result);
            expect(result).toBeDefined();
        });
        it('should not throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, props01, 'prop', SongList.name);
            expect(result).toBeDefined();
        });
        it('should throw a warning', ()=>{
            const result = checkPropTypes(SongList.propTypes, props00, 'prop', SongList.name);
            console.log(result);
            expect(result).toBeDefined();
        });
    });
});
