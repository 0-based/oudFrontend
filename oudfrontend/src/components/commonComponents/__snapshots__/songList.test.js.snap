// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongList Component snapshot test renders correctly 1`] = `
<div
  className="col-xs-8 col-md-6 col-lg-8 col-xl-8"
  data-testid="songsList"
>
  <button
    className="song row"
    data-testid="song"
    id="song"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="songIcon col-1"
    >
      <button
        className="songButton"
        data-testid="playButton"
        onClick={[Function]}
      >
        <img
          alt="play music icon"
          data-testid="playButtonImage"
          height="14"
          src="musicIcon.png"
          width="12"
        />
      </button>
    </div>
    <div
      className="songInfo col-8"
    >
      <span
        className="whiteText"
        data-testid="songName"
      >
        إنسان
      </span>
      <p
        data-testid="aristsNames"
      >
        <span>
          <span>
            <a
              className="playlistAnchor"
              data-testid="artistName"
              href="/artist/5"
              onClick={[Function]}
            >
              حمزة نمرة
            </a>
            <span
              data-testid="comma"
            >
              , 
            </span>
          </span>
          <span>
            <a
              className="playlistAnchor"
              data-testid="artistName"
              href="/artist/2"
              onClick={[Function]}
            >
              Sardeena
            </a>
            <span
              data-testid="comma"
            >
              , 
            </span>
          </span>
        </span>
        <a
          className="playlistAnchor"
          data-testid="albumName"
          href="/albums/1"
          onClick={[Function]}
        >
          
        </a>
      </p>
    </div>
    <div
      className="col-1"
    >
      <div
        className="dropdown"
        data-testid="dropdown"
      >
        <button
          className="songButton"
          data-testid="dropdownButton"
          id="songDropdownButton"
          onClick={[Function]}
        >
          <h3
            className="whiteText"
            style={
              Object {
                "display": "none",
              }
            }
          >
            ...
          </h3>
        </button>
        <div
          className="dropdownContent"
          data-testid="dropdownList"
          id="dropdownContent"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <a
            className="SongDropdownItem"
            data-testid="saveSong"
            href=""
            onClick={[Function]}
          >
            Save to your Liked Songs
          </a>
          <a
            className="SongDropdownItem"
            data-testid="addToQueue"
            href=""
            onClick={[Function]}
          >
            Add to Queue
          </a>
          <a
            className="SongDropdownItem"
            data-testid="addToPlaylist"
            href=""
            onClick={[Function]}
          >
            Add to Playlist
          </a>
        </div>
      </div>
    </div>
    <div
      className="col-2"
    >
      <p
        className="whiteText"
        data-testid="songTime"
      >
        3:34
      </p>
    </div>
  </button>
  <button
    className="song row"
    data-testid="song"
    id="song"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="songIcon col-1"
    >
      <button
        className="songButton"
        data-testid="playButton"
        onClick={[Function]}
      >
        <img
          alt="play music icon"
          data-testid="playButtonImage"
          height="14"
          src="musicIcon.png"
          width="12"
        />
      </button>
    </div>
    <div
      className="songInfo col-8"
    >
      <span
        className="whiteText"
        data-testid="songName"
      >
        سلاح التلميذ
      </span>
      <p
        data-testid="aristsNames"
      >
        <span>
          <span>
            <a
              className="playlistAnchor"
              data-testid="artistName"
              href="/artist/3"
              onClick={[Function]}
            >
              ًWegz el wench
            </a>
            <span
              data-testid="comma"
            >
              , 
            </span>
          </span>
          <span>
            <a
              className="playlistAnchor"
              data-testid="artistName"
              href="/artist/4"
              onClick={[Function]}
            >
              Amr Diab
            </a>
            <span
              data-testid="comma"
            >
              , 
            </span>
          </span>
        </span>
        <a
          className="playlistAnchor"
          data-testid="albumName"
          href="/albums/1"
          onClick={[Function]}
        >
          
        </a>
      </p>
    </div>
    <div
      className="col-1"
    >
      <div
        className="dropdown"
        data-testid="dropdown"
      >
        <button
          className="songButton"
          data-testid="dropdownButton"
          id="songDropdownButton"
          onClick={[Function]}
        >
          <h3
            className="whiteText"
            style={
              Object {
                "display": "none",
              }
            }
          >
            ...
          </h3>
        </button>
        <div
          className="dropdownContent"
          data-testid="dropdownList"
          id="dropdownContent"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <a
            className="SongDropdownItem"
            data-testid="saveSong"
            href=""
            onClick={[Function]}
          >
            Save to your Liked Songs
          </a>
          <a
            className="SongDropdownItem"
            data-testid="addToQueue"
            href=""
            onClick={[Function]}
          >
            Add to Queue
          </a>
          <a
            className="SongDropdownItem"
            data-testid="addToPlaylist"
            href=""
            onClick={[Function]}
          >
            Add to Playlist
          </a>
        </div>
      </div>
    </div>
    <div
      className="col-2"
    >
      <p
        className="whiteText"
        data-testid="songTime"
      >
        3:34
      </p>
    </div>
  </button>
</div>
`;
